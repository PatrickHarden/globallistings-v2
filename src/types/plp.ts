export interface PlpQuery {
  readonly aspects?: string;
  readonly sort?: string;
  readonly radius?: string;
}

export enum TransactionType {
  sale = 'sale',
  letting = 'letting',
}

export interface PlpPath {
  readonly usageType: string;
  readonly transactionType: TransactionType;
  readonly location: string;
}

export function isPlpPathValue(value: string): value is keyof PlpPath {
  if (value === 'usageType' || value === 'transactionType' || value === 'location') {
    return true;
  }

  return false;
}

export function isPlpPathObject(path: object | null): path is PlpPath {
  const p = path as PlpPath;
  if (!p) {
    return false;
  }
  return !!p.usageType && !!p.transactionType && !!p.location;
}
