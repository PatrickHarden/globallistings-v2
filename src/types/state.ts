import { RouterState } from 'connected-react-router';
import { Dictionary } from './common/dictionary';
import { Property } from './property';
import { Config, AppConfig } from './config';
import { UI } from './ui';
import { Autocomplete } from './autocomplete';

export interface State {
  readonly entities: Entities;
  readonly entitiesMeta: EntitiesMeta;
  readonly router: RouterState;
  readonly ui: UI;
}

export interface EntitiesMeta {
  readonly propertySearches: Dictionary<EntityMeta>;
  readonly geocode: Dictionary<EntityMeta>;
  readonly autocompleteSearches: Dictionary<EntityMeta>;
  readonly config: EntityMeta;
}

export type LoadStatus = 'PENDING' | 'ERROR' | 'SUCCESS' | 'NONE';

export interface EntityMeta {
  readonly status: LoadStatus;
}

export interface LatLng {
  readonly lat: number;
  readonly lng: number;
}

export interface Bounds {
  readonly sw: LatLng;
  readonly ne: LatLng;
}

export interface Geocode {
  readonly bounds: Bounds;
  readonly center: LatLng;
}

export interface Entities {
  readonly properties: Dictionary<Property>;
  readonly propertySearches: Dictionary<string[]>;
  readonly autocomplete: Dictionary<Autocomplete>;
  readonly autocompleteSearches: string[];
  readonly geocode: Dictionary<Geocode>;
  readonly config: AppConfig;
}

export function isConfigObject(config: string | Config): config is Config {
  return typeof config !== 'string';
}
