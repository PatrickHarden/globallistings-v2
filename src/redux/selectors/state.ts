import { State } from '../../types/state';
import { theme } from '../../themes/default/theme';
import { createSelector } from 'reselect';
import { FilterType } from '../../types/config';

// State
export const entityMetaSelector = (state: State) => state.entitiesMeta;
export const entitiesSelector = (state: State) => state.entities;
export const uiSelector = (state: State) => state.ui;
export const locationSelector = (state: State) => state.router.location;

// Entities
export const propertiesSelector = createSelector(
  entitiesSelector,
  entities => entities.properties
);

export const geocodeSelector = createSelector(
  entitiesSelector,
  entities => entities.geocode
);

export const propertySearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.propertySearches
);

export const configSelector = createSelector(
  entitiesSelector,
  entities => entities.config
);

export const autocompleteSelector = createSelector(
  entitiesSelector,
  entities => entities.autocomplete
);

export const autocompleteSearchesSelector = createSelector(
  entitiesSelector,
  entities => entities.autocompleteSearches
);

// Meta
export const propertiesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.propertySearches
);

export const geocodeMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.geocode
);

export const configMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.config
);

export const autocompleteSearchesMetaSelector = createSelector(
  entityMetaSelector,
  meta => meta.autocompleteSearches
);

// Config
export const localeSelector = createSelector(
  configSelector,
  config => config.language
);

export const messagesSelector = createSelector(
  configSelector,
  config => config.i18n
);

export const filtersSelector = createSelector(
  configSelector,
  config => config.filters
);

export const rangeFiltersSelector = createSelector(
  filtersSelector,
  filters => filters.filter(f => f.type === FilterType.range)
);

// Other

// TODO add logic to switch between themes based on config
export const themeSelector = (state: State) => theme;

// TODO figure out where it will be coming from
export const googleMapsApiKeySelector = () => 'AIzaSyCzEw2NaQFixR7sZCLnZ1oGQljW7AxLZ7k';
