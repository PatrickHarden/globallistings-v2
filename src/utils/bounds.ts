import { Bounds } from '../types/state';

export const googleBoundsToBounds = (bounds: google.maps.LatLngBounds): Bounds => {
  const ne = bounds.getNorthEast();
  const sw = bounds.getSouthWest();

  return {
    ne: {
      lat: ne.lat(),
      lng: ne.lng(),
    },
    sw: {
      lat: sw.lat(),
      lng: sw.lng(),
    },
  };
};

export const boundsToGoogleBounds = (bounds: Bounds): google.maps.LatLngBounds =>
  new google.maps.LatLngBounds(bounds.sw, bounds.ne);
