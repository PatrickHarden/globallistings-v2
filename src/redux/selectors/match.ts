import { createMatchSelector, RouterRootState } from 'connected-react-router';
import { PlpPath } from '../../types/plp';
import { PdpPath } from '../../types/pdp';
import RoutePaths from '../../constants/routePaths';
import { createSelector } from 'reselect';

export const plpMatchSelector = createMatchSelector<RouterRootState, PlpPath>(RoutePaths.plp);
export const pdpMatchSelector = createMatchSelector<RouterRootState, PdpPath>(RoutePaths.pdp);

export const pathParamsSelector = createSelector(
  plpMatchSelector,
  pdpMatchSelector,
  (plpMatch, pdpMatch) => ({ ...(plpMatch || {}), ...(pdpMatch || {}) })
);

export const plpUrlPathParamsSelector = createSelector(
  plpMatchSelector,
  match => {
    if (!match) {
      throw new Error('Url path must be defined to use this selector');
    }
    return match.params;
  }
);

export const pdpUrlPathParamsSelector = createSelector(
  pdpMatchSelector,
  match => {
    if (!match) {
      throw new Error('Url path must be defined to use this selector');
    }
    return match.params;
  }
);
