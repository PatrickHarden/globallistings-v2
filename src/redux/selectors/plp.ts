import { createSelector } from 'reselect';
import {
  geocodeMetaSelector,
  propertiesMetaSelector,
  propertySearchesSelector,
  configSelector,
} from './state';
import { propertiesSearchRequestQuerySelector } from './search-api-request';
import { murmur2 } from 'murmurhash-js';
import { plpLocationNameSelector } from './location';
import { otherConfigsSelector } from './config/config';
import { plpUrlPathParamsSelector, plpMatchSelector } from './match';

export const plpQueryHashSelector = createSelector(
  propertiesSearchRequestQuerySelector,
  query => murmur2(JSON.stringify(query)).toString()
);

export const plpPropertyRequestStatus = createSelector(
  plpQueryHashSelector,
  propertiesMetaSelector,
  (hash, propertiesMeta) => {
    const propertyMeta = propertiesMeta[hash];
    if (propertyMeta === undefined) {
      return 'NONE';
    }

    return propertyMeta.status;
  }
);

export const propertySearchSelector = createSelector(
  propertySearchesSelector,
  plpQueryHashSelector,
  (propertySearches, hash) => {
    const result = propertySearches[hash];

    if (result) {
      return result;
    }

    return [];
  }
);

const plpLocationGeocodeStatus = createSelector(
  plpLocationNameSelector,
  geocodeMetaSelector,
  (location, meta) => {
    const entityMeta = meta[location];
    return entityMeta ? entityMeta.status : 'NONE';
  }
);

export const isPlpLocationGeocoded = createSelector(
  plpLocationGeocodeStatus,
  status => status === 'SUCCESS'
);

export const isUsageTypeConfigLoaded = createSelector(
  configSelector,
  plpUrlPathParamsSelector,
  (config, path) => {
    if (!config.currentConfig) {
      return true;
    }

    return path.usageType === config.currentConfig;
  }
);

export const usageTypeSelector = createSelector(
  plpUrlPathParamsSelector,
  params => params.usageType
);

export const configEndpointForCurrentUsageTypeSelector = createSelector(
  usageTypeSelector,
  otherConfigsSelector,
  (usageType, configs) => {
    const config = configs.find(c => c.name === usageType);
    if (!config) {
      throw new Error('Config is missing data for config endpoint mapping');
    }

    return config.endpoint;
  }
);

export const isPlpPath = createSelector(
  plpMatchSelector,
  match => !!match
);
