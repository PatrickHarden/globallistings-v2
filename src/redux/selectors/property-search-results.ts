import { createSelector } from 'reselect';
import { InjectedIntl } from 'react-intl';
import { propertiesSelector } from './state';
import { intlSelector } from './intl';
import { notEmpty } from '../../types/common/notEmpty';
import { Location, Property } from '../../types/property';
import { propertySearchSelector, usageTypeSelector } from './plp';
import { selectedPropertyPinSelector, highlightedPropertySelector } from './ui';
import { plpTransactionTypeSelector } from './location';

import { mapAddress } from '../../utils/propertyData/mapAddress';
import { mapImage } from '../../utils/propertyData/mapImage';
import { mapPrice } from '../../utils/propertyData/mapPrice';
import { mapSize } from '../../utils/propertyData/mapSize';
import { capitalise } from '../../utils/capitalise';

export interface PropertyOverview {
  readonly imgUrl: string;
  readonly price?: string;
  readonly size?: string;
  readonly address: string;
  readonly location: Location;
  readonly id: string;
  readonly selected: boolean;
}

// TODO update to a real one
export const isCommercialSelector = () => true;

const toPropertyOverview = (
  intl: InjectedIntl,
  isCommercial: boolean,
  selectedPropertyId: string,
  highlightedPropertyId: string
) => (p: Property, i: Number): PropertyOverview => ({
  address: mapAddress(p, intl),
  imgUrl: mapImage(p),
  price: mapPrice(p, intl),
  size: mapSize(p, intl, isCommercial),
  location: p.location,
  id: p.id,
  // Selected poperty = currently highlighted (mouseover list), selected (pin clicked) or first in array
  selected: highlightedPropertyId
    ? highlightedPropertyId === p.id
    : selectedPropertyId === p.id || i === 0,
});

export const propertiesSearchResultsSelector = createSelector(
  propertiesSelector,
  propertySearchSelector,
  intlSelector,
  isCommercialSelector,
  selectedPropertyPinSelector,
  highlightedPropertySelector,
  (properties, propertySearch, intl, isCommercial, selectedPropertyId, highlightedPropertyId) => {
    return propertySearch
      .map(p => properties[p])
      .filter(notEmpty)
      .map(toPropertyOverview(intl, isCommercial, selectedPropertyId, highlightedPropertyId));
  }
);

export const resultCountSelector = createSelector(
  propertiesSearchResultsSelector,
  usageTypeSelector,
  plpTransactionTypeSelector,
  intlSelector,
  (results, usage, transaction, { formatMessage }) =>
    formatMessage(
      { id: 'PropertiesFound' },
      {
        propertyCount: results.length,
        propertyTypePlural: formatMessage({
          id: `PDPPluralPropertyType${capitalise(usage)}`,
        }).toLocaleLowerCase(),
        searchType: formatMessage({
          id: transaction === 'sale' ? 'saleSearchType' : 'letSearchType',
        }),
      }
    )
);
