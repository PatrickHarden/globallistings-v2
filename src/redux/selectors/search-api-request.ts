// @ts-nocheck
import { PlpQuery, TransactionType } from '../../types/plp';
import { createSelector } from 'reselect';
import renameKeys from '../../utils/renameKeys';
import { paramsSelector } from './config/config';
import { querySelector, plpTransactionTypeSelector } from './location';
import { configSelector, filtersSelector } from './state';
import { boundsToSquare } from '../../utils/boundsToSquare';
import { areaSelector } from './area';
import { TransactionTypeValue } from '../../types/config';
import { filter, is, Dictionary } from 'ramda';
import { filterParamsCombiner } from './config/range-filters';

const QueryToApiRequestParamsMapping: { [P in keyof PlpQuery]?: string } = {
  sort: 'Sort',
};

const SelectProperties =
  'Dynamic.PrimaryImage,Common.ActualAddress,Common.Charges,Common.NumberOfBedrooms,Common.PrimaryKey,Common.Coordinate,Common.Aspects,Common.ListingCount,Common.IsParent,Common.ContactGroup,Common.Highlights,Common.Walkthrough,Common.MinimumSize,Common.MaximumSize,Common.TotalSize,Common.GeoLocation,Common.Sizes';

const mappedApiRequestQuery = createSelector(querySelector, query =>
  renameKeys(QueryToApiRequestParamsMapping, query)
);

const locationParamsSelector = createSelector(areaSelector, area => {
  switch (area.type) {
    case 'circle':
      return {
        radius: area.radius.toString(),
        Lat: area.center.lat,
        Lon: area.center.lng,
      };
    case 'rectangle':
      return {
        PolygonFilters: JSON.stringify([boundsToSquare(area.bounds)]),
      };
  }
});

export const paginationSelector = createSelector(configSelector, config => ({
  PageSize: config.limitListMapResults ? Number(config.limitListMapResults) : undefined,
  Page: 1, // TODO: Hook up to pagination functionality
}));

const transactionTypeMapping: { [K in TransactionType]: TransactionTypeValue | undefined } = {
  sale: TransactionTypeValue.isSale,
  letting: TransactionTypeValue.isLetting,
};

export const transactionTypeParamsSelector = createSelector(
  plpTransactionTypeSelector,
  transactionType => {
    const result = transactionTypeMapping[transactionType];
    if (!result) {
      throw new Error(`${transactionType} is not a supported transaction type`);
    }

    return { 'Common.Aspects': result };
  }
);

// NOTE: we have to cast to the dictionary, as the typing can't infer that undefined will be filters out with is(String)
const queryParamsSelector = createSelector(
  querySelector,
  query => filter(is(String), query) as Dictionary<string>
);

const rangeFilterParamsSelector = createSelector(
  queryParamsSelector,
  filtersSelector,
  filterParamsCombiner
);

export const propertiesSearchRequestQuerySelector = createSelector(
  mappedApiRequestQuery,
  locationParamsSelector,
  paramsSelector,
  paginationSelector,
  transactionTypeParamsSelector,
  rangeFilterParamsSelector,
  (query, locationParams, defaultParams, pagination, transactionTypeParams, rangeFilterParams) => ({
    ...pagination,
    ...defaultParams,
    ...locationParams,
    ...query,
    ...transactionTypeParams,
    ...rangeFilterParams,
    // _select: SelectProperties,
  })
);

export const propertySearchRequestQuerySelector = createSelector(paramsSelector, defaultParams => ({
  ...defaultParams,
  _select: SelectProperties,
}));
