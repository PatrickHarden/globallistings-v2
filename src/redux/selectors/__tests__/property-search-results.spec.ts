jest.mock('murmurhash-js');
jest.mock('../match');
jest.mock('../../../utils/bounds');

import { DeepPartial } from '../../../types/common/deep-partial';
import { State } from '../../../types/state';
import { murmur2 } from 'murmurhash-js';
import { propertiesSearchResultsSelector, resultCountSelector } from '../property-search-results';
import i18nJson from '../../../tests/fixtures/store-i18n.json';
import { Dictionary } from '../../../types/common/dictionary';
import { plpQueryHashSelector, plpPropertyRequestStatus } from '../plp';
import { propertiesSearchRequestQuerySelector } from '../search-api-request';
import { googleBoundsToBounds } from '../../../utils/bounds';
import { plpUrlPathParamsSelector } from '../match';

const i18n = i18nJson as Dictionary<string>;
const MOCK_HASH = 'MOCK_HASH';
const murmurMock = murmur2 as jest.Mock;

const baseState: DeepPartial<State> = {
  entities: {
    config: {
      limitListMapResults: '100',
      params: {
        Sort: 'bam',
      },
      filters: [],
    },
    geocode: {
      london: {
        bounds: {
          ne: { lat: 1, lng: 2 },
          sw: { lat: 1, lng: 2 },
        },
        center: {
          lat: 1,
          lng: 2,
        },
      },
    },
  },
};

(googleBoundsToBounds as jest.Mock).mockReturnValue({
  ne: { lat: 1, lng: 1 },
  sw: { lat: 2, lng: 2 },
});

beforeEach(() => {
  ((plpUrlPathParamsSelector as unknown) as jest.Mock).mockReturnValue({
    location: 'london',
    transactionType: 'sale',
    usageType: 'office',
  });
  (murmur2 as jest.Mock).mockReturnValue(MOCK_HASH);
});

describe(plpQueryHashSelector, () => {
  afterEach(() => {
    jest.clearAllMocks();
  });

  it('should merge config with query params', () => {
    // Arrange
    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?radius=10' } },
    };

    // Act
    const result = plpQueryHashSelector(state as State);

    // Assert
    expect(result).toEqual(MOCK_HASH);
    expect(murmur2).toHaveBeenCalledWith(
      '{"PageSize":100,"Page":1,"Sort":"bam","radius":"10","Lat":1,"Lon":2,"Common.Aspects":"isSale"}'
    );
  });

  it('query should overwrite default config', () => {
    // Arrange
    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?sort=zap' } },
    };

    // Act
    const result = plpQueryHashSelector(state as State);

    // Assert
    expect(murmur2).toHaveBeenCalledWith(
      '{"PageSize":100,"Page":1,"Sort":"zap","PolygonFilters":"[[\\"1,2\\",\\"1,2\\",\\"1,2\\",\\"1,2\\"]]","Common.Aspects":"isSale"}'
    );
    expect(result).toEqual(MOCK_HASH);
  });

  it('query params should not be affected by order', () => {
    // Arrange
    const state1: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?Site=uk&PageSize=1' } },
    };
    const state2: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?PageSize=1&Site=uk' } },
    };

    // Act
    plpQueryHashSelector(state1 as State);
    plpQueryHashSelector(state2 as State);

    // Assert
    expect(murmurMock.mock.calls[0][0]).toEqual(murmurMock.mock.calls[1][0]);
  });
});

describe(propertiesSearchRequestQuerySelector, () => {
  it('should generate a key for query param', () => {
    // Arrange
    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?radius=5' } },
    };

    // Act
    const result = propertiesSearchRequestQuerySelector(state as State);

    // Assert
    expect(result).toEqual({
      'Common.Aspects': 'isSale',
      Lat: 1,
      Lon: 2,
      Page: 1,
      PageSize: 100,
      Sort: 'bam',
      radius: '5',
    });
  });

  it('should generate a key for query param and path for sale', () => {
    // Arrange
    ((plpUrlPathParamsSelector as unknown) as jest.Mock).mockReturnValue({
      transactionType: 'sale',
      location: 'london',
    });
    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?radius=5' } },
    };

    // Act
    const result = propertiesSearchRequestQuerySelector(state as State);

    // Assert
    expect(result).toEqual({
      'Common.Aspects': 'isSale',
      PageSize: 100,
      Page: 1,
      Lat: 1,
      Lon: 2,
      Sort: 'bam',
      radius: '5',
    });
  });

  it('should generate a key for query param and path for let', () => {
    // Arrange
    ((plpUrlPathParamsSelector as unknown) as jest.Mock).mockReturnValue({
      transactionType: 'letting',
      location: 'london',
    });

    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?radius=5' } },
    };

    // Act
    const result = propertiesSearchRequestQuerySelector(state as State);

    // Assert
    expect(result).toEqual({
      'Common.Aspects': 'isLetting',
      PageSize: 100,
      Page: 1,
      Lat: 1,
      Lon: 2,
      Sort: 'bam',
      radius: '5',
    });
  });

  it('should throw an exception for unrecognised usageType', () => {
    // Arrange
    ((plpUrlPathParamsSelector as unknown) as jest.Mock).mockReturnValue({
      transactionType: 'zap',
      location: 'london',
    });

    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?radius=5' } },
    };

    // Assert
    expect(() => propertiesSearchRequestQuerySelector(state as State)).toThrowError();
  });
});

describe(plpPropertyRequestStatus, () => {
  it('should return NONE if no request has been made', () => {
    // Arrange
    const state: DeepPartial<State> = {
      ...baseState,
      router: { location: { search: '?Site=uk' } },
      entitiesMeta: {
        propertySearches: {
          [MOCK_HASH]: {
            status: 'PENDING',
          },
        },
      },
    };

    // Act
    const result = plpPropertyRequestStatus(state as State);

    // Assert
    expect(result).toEqual('PENDING');
  });
});

describe(propertiesSearchResultsSelector, () => {
  // Arrange
  const photos = {
    hero: {
      large: 'mock_url',
      small: 'mock_url',
    },
  };
  const state: DeepPartial<State> = {
    router: { location: { search: '?Site=uk' } },

    entities: {
      ...baseState.entities,
      config: {
        i18n,
        language: 'en-GB',
        filters: [],
      },
      properties: {
        P1: {
          address: {
            Line1: 'London Str 12',
            Line2: 'Block A',
          },
          photos,
          charges: [],
        },
        P2: {
          photos,
          address: {
            Line1: 'London Str 12',
            Line2: 'Block A',
            PostCode: 'N123FD',
            Locality: 'London',
          },
          size: {
            min: {
              amount: 200,
              units: 'sqft',
            },
            max: {
              amount: 200,
              units: 'sqft',
            },
          },
          charges: [],
        },
        P3: {
          photos,
          address: {
            Line1: 'London Str 12',
            Line2: 'Block A',
            PostCode: 'N123FD',
            Locality: 'London',
          },
          size: {
            min: {
              amount: 200,
              units: 'sqft',
            },
            max: {
              amount: 3000,
              units: 'sqft',
            },
          },
          charges: [
            {
              chargeKind: 'Rent',
              amount: 200,
              perUnit: 'sqft',
              interval: 'year',
            },
          ],
        },
      },
      propertySearches: {
        [MOCK_HASH]: ['P1', 'P2', 'P3'],
      },
    },
    ui: {
      map: {},
      list: {},
    },
  };

  it('should return correct data', () => {
    const result = propertiesSearchResultsSelector(state as State);
    expect(result).toMatchSnapshot();
  });

  it('should return correct data with a selected property', () => {
    const newState = {
      ...state,
      ui: {
        ...state.ui,
        map: {
          selectedPropertyPin: 'P1',
        },
      },
    };
    const result = propertiesSearchResultsSelector(newState as State);
    expect(result).toMatchSnapshot();
  });

  it('should return correct data with a highlighted property', () => {
    const newState = {
      ...state,
      ui: {
        ...state.ui,
        list: {
          highlightedProperty: 'P2',
        },
      },
    };
    const result = propertiesSearchResultsSelector(newState as State);
    expect(result).toMatchSnapshot();
  });
});

describe(resultCountSelector, () => {
  it('should return correctly formatted string', () => {
    const state: DeepPartial<State> = {
      entities: {
        config: {
          filters: [],
          language: 'en',
          i18n: {
            PDPPluralPropertyTypeOffice: 'Offices',
            PropertiesFound: '{propertyCount} matching {propertyTypePlural} for {searchType}',
            saleSearchType: 'sale',
          },
        },
        geocode: {
          london: {
            bounds: {
              ne: { lat: 1, lng: 2 },
              sw: { lat: 1, lng: 2 },
            },
            center: {
              lat: 1,
              lng: 2,
            },
          },
        },
        properties: {
          P1: {
            address: {
              Line1: 'London Str 1',
            },
            photos: {
              hero: {
                large: 'mock_url',
                small: 'mock_url',
              },
            },
          },
          P2: {
            address: {
              Line1: 'London Str 2',
            },
            photos: {
              hero: {
                large: 'mock_url',
                small: 'mock_url',
              },
            },
          },
        },
        propertySearches: {
          [MOCK_HASH]: ['P1', 'P2'],
        },
      },
      ui: {
        map: {},
        list: {},
      },
      router: {
        action: 'PUSH',
        location: {
          pathname: '/office/sale/london/results',
          search: '',
        },
      },
    };

    const result = resultCountSelector(state as State);

    expect(result).toEqual('2 matching offices for sale');
  });
});
