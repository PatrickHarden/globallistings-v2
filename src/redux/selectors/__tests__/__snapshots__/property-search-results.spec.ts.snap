// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should return correct data 1`] = `
Array [
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": true,
    "size": undefined,
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": true,
    "size": "200sqft",
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "200/sqft/year",
    "selected": true,
    "size": "200sqft - 3000sqft",
  },
]
`;

exports[`should return correct data with a highlighted property 1`] = `
Array [
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": false,
    "size": undefined,
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": false,
    "size": "200sqft",
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "200/sqft/year",
    "selected": false,
    "size": "200sqft - 3000sqft",
  },
]
`;

exports[`should return correct data with a selected property 1`] = `
Array [
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": true,
    "size": undefined,
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "Price On Application",
    "selected": false,
    "size": "200sqft",
  },
  Object {
    "address": "London Str 12 Block A",
    "id": undefined,
    "imgUrl": "mock_url",
    "location": undefined,
    "price": "200/sqft/year",
    "selected": false,
    "size": "200sqft - 3000sqft",
  },
]
`;
